File: suc/kotlin/078.kt
package com.example.myapplication

import android.os.Bundle
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.LazyRow
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.CheckCircle
import androidx.compose.material.icons.filled.Face
import androidx.compose.material.icons.filled.Favorite
import androidx.compose.material.icons.filled.Home
import androidx.compose.material.icons.filled.Notifications
import androidx.compose.material.icons.filled.PlayArrow
import androidx.compose.material.icons.filled.Search
import androidx.compose.material.icons.filled.Settings
import androidx.compose.material.icons.filled.Star
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.Icon
import androidx.compose.material3.IconButton
import androidx.compose.material3.LinearProgressIndicator
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.material3.lightColorScheme
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.core.view.WindowCompat
import androidx.core.view.WindowInsetsCompat
import androidx.core.view.WindowInsetsControllerCompat

class MainActivity : ComponentActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val windowInsetsController = WindowCompat.getInsetsController(window, window.decorView)
        windowInsetsController.systemBarsBehavior =
            WindowInsetsControllerCompat.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE
        windowInsetsController.hide(WindowInsetsCompat.Type.systemBars())
        WindowCompat.setDecorFitsSystemWindows(window, false)

        setContent {
            AppTheme {
                Surface(
                    modifier = Modifier.fillMaxSize(),
                    color = MaterialTheme.colorScheme.background
                ) {
                    CourseDashboardScreen()
                }
            }
        }
    }
}

@Composable
fun AppTheme(content: @Composable () -> Unit) {
    val colorScheme = lightColorScheme(
        primary = Color(0xFF4E31AA),
        secondary = Color(0xFF3A1C71),
        background = Color(0xFFF5F5F7),
        surface = Color.White,
        onPrimary = Color.White,
        onBackground = Color(0xFF1C1B1F),
        onSurface = Color(0xFF1C1B1F)
    )
    MaterialTheme(
        colorScheme = colorScheme,
        content = content
    )
}

@Composable
fun CourseDashboardScreen() {
    Box(modifier = Modifier.fillMaxSize()) {
        Column(
            modifier = Modifier
                .fillMaxSize()
                .verticalScroll(rememberScrollState())
                .padding(bottom = 80.dp)
        ) {
            HeaderSection()
            SearchBarSection()
            OngoingCourseSection()
            CategoriesSection()
            RecommendedSection()
        }

        BottomNavigationBar(
            modifier = Modifier.align(Alignment.BottomCenter)
        )
    }
}

@Composable
fun HeaderSection() {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .padding(24.dp),
        horizontalArrangement = Arrangement.SpaceBetween,
        verticalAlignment = Alignment.CenterVertically
    ) {
        Column {
            Text(
                text = "Hello, Sophie",
                style = MaterialTheme.typography.headlineMedium,
                fontWeight = FontWeight.Bold
            )
            Text(
                text = "Let's start learning!",
                style = MaterialTheme.typography.bodyLarge,
                color = Color.Gray
            )
        }
        Row {
            IconButton(onClick = { }) {
                Icon(
                    imageVector = Icons.Default.Notifications,
                    contentDescription = null,
                    tint = MaterialTheme.colorScheme.onBackground
                )
            }
            Spacer(modifier = Modifier.width(8.dp))
            Box(
                modifier = Modifier
                    .size(48.dp)
                    .clip(CircleShape)
                    .background(MaterialTheme.colorScheme.primary.copy(alpha = 0.2f)),
                contentAlignment = Alignment.Center
            ) {
                Icon(
                    imageVector = Icons.Default.Face,
                    contentDescription = null,
                    tint = MaterialTheme.colorScheme.primary
                )
            }
        }
    }
}

@Composable
fun SearchBarSection() {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .padding(horizontal = 24.dp)
            .height(56.dp)
            .clip(RoundedCornerShape(16.dp))
            .background(Color.White)
            .padding(horizontal = 16.dp),
        verticalAlignment = Alignment.CenterVertically
    ) {
        Icon(
            imageVector = Icons.Default.Search,
            contentDescription = null,
            tint = Color.Gray
        )
        Spacer(modifier = Modifier.width(16.dp))
        Text(
            text = "Search for course...",
            color = Color.LightGray
        )
        Spacer(modifier = Modifier.weight(1f))
        Icon(
            imageVector = Icons.Default.Settings,
            contentDescription = null,
            tint = MaterialTheme.colorScheme.primary
        )
    }
}

@Composable
fun OngoingCourseSection() {
    Column(modifier = Modifier.padding(24.dp)) {
        Text(
            text = "Ongoing Course",
            style = MaterialTheme.typography.titleLarge,
            fontWeight = FontWeight.Bold
        )
        Spacer(modifier = Modifier.height(16.dp))
        Card(
            modifier = Modifier
                .fillMaxWidth()
                .height(140.dp),
            shape = RoundedCornerShape(24.dp),
            colors = CardDefaults.cardColors(containerColor = MaterialTheme.colorScheme.primary)
        ) {
            Row(
                modifier = Modifier
                    .fillMaxSize()
                    .padding(20.dp),
                verticalAlignment = Alignment.CenterVertically
            ) {
                Column(modifier = Modifier.weight(1f)) {
                    Text(
                        text = "3D Modeling",
                        color = Color.White.copy(alpha = 0.7f),
                        fontSize = 14.sp
                    )
                    Spacer(modifier = Modifier.height(8.dp))
                    Text(
                        text = "Blender Basic",
                        color = Color.White,
                        fontSize = 22.sp,
                        fontWeight = FontWeight.Bold
                    )
                    Spacer(modifier = Modifier.height(16.dp))
                    Row(verticalAlignment = Alignment.CenterVertically) {
                        Text(text = "75%", color = Color.White, fontSize = 12.sp)
                        Spacer(modifier = Modifier.width(8.dp))
                        LinearProgressIndicator(
                            progress = 0.75f,
                            modifier = Modifier
                                .weight(1f)
                                .height(6.dp)
                                .clip(RoundedCornerShape(4.dp)),
                            color = Color(0xFFFFA726),
                            trackColor = Color.White.copy(alpha = 0.2f)
                        )
                    }
                }
                Spacer(modifier = Modifier.width(16.dp))
                Box(
                    modifier = Modifier
                        .size(60.dp)
                        .clip(CircleShape)
                        .background(Color.White),
                    contentAlignment = Alignment.Center
                ) {
                    Icon(
                        imageVector = Icons.Default.PlayArrow,
                        contentDescription = null,
                        tint = MaterialTheme.colorScheme.primary,
                        modifier = Modifier.size(32.dp)
                    )
                }
            }
        }
    }
}

@Composable
fun CategoriesSection() {
    val categories = listOf("Coding", "Design", "Business", "Photo", "Music")
    Column {
        PaddingValues(horizontal = 24.dp)
        LazyRow(
            contentPadding = PaddingValues(horizontal = 24.dp),
            horizontalArrangement = Arrangement.spacedBy(12.dp)
        ) {
            items(categories) { category ->
                val isSelected = category == "Design"
                Box(
                    modifier = Modifier
                        .clip(RoundedCornerShape(12.dp))
                        .background(if (isSelected) Color(0xFFFF7043) else Color.White)
                        .padding(horizontal = 24.dp, vertical = 12.dp)
                ) {
                    Text(
                        text = category,
                        color = if (isSelected) Color.White else Color.Gray,
                        fontWeight = FontWeight.Medium
                    )
                }
            }
        }
    }
}

@Composable
fun RecommendedSection() {
    val courses = listOf(
        Course("Mobile Dev", "Android Compose", "4.9", Color(0xFFE3F2FD), Color(0xFF1565C0)),
        Course("Photography", "Master Lighting", "4.7", Color(0xFFFCE4EC), Color(0xFFC2185B)),
        Course("Finance", "Crypto Trading", "4.5", Color(0xFFE8F5E9), Color(0xFF2E7D32))
    )

    Column(modifier = Modifier.padding(24.dp)) {
        Row(
            modifier = Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.SpaceBetween,
            verticalAlignment = Alignment.CenterVertically
        ) {
            Text(
                text = "Recommended",
                style = MaterialTheme.typography.titleLarge,
                fontWeight = FontWeight.Bold
            )
            Text(
                text = "See all",
                color = Color.Gray
            )
        }
        Spacer(modifier = Modifier.height(16.dp))

        courses.forEach { course ->
            CourseCard(course)
            Spacer(modifier = Modifier.height(16.dp))
        }
    }
}

@Composable
fun CourseCard(course: Course) {
    Card(
        modifier = Modifier
            .fillMaxWidth()
            .height(100.dp),
        shape = RoundedCornerShape(20.dp),
        colors = CardDefaults.cardColors(containerColor = Color.White)
    ) {
        Row(
            modifier = Modifier.fillMaxSize(),
            verticalAlignment = Alignment.CenterVertically
        ) {
            Box(
                modifier = Modifier
                    .fillMaxHeight()
                    .width(100.dp)
                    .background(course.bgColor),
                contentAlignment = Alignment.Center
            ) {
                Icon(
                    imageVector = if (course.title.contains("Mobile")) Icons.Default.CheckCircle else Icons.Default.Star,
                    contentDescription = null,
                    tint = course.accentColor,
                    modifier = Modifier.size(32.dp)
                )
            }
            Column(
                modifier = Modifier
                    .weight(1f)
                    .padding(16.dp)
            ) {
                Text(
                    text = course.subtitle,
                    color = MaterialTheme.colorScheme.primary,
                    fontSize = 12.sp,
                    fontWeight = FontWeight.Bold
                )
                Text(
                    text = course.title,
                    style = MaterialTheme.typography.titleMedium,
                    fontWeight = FontWeight.Bold
                )
                Spacer(modifier = Modifier.height(4.dp))
                Row(verticalAlignment = Alignment.CenterVertically) {
                    Icon(
                        imageVector = Icons.Default.Star,
                        contentDescription = null,
                        tint = Color(0xFFFFB300),
                        modifier = Modifier.size(16.dp)
                    )
                    Spacer(modifier = Modifier.width(4.dp))
                    Text(
                        text = course.rating,
                        color = Color.Gray,
                        fontSize = 12.sp
                    )
                }
            }
            IconButton(onClick = { }) {
                Icon(
                    imageVector = Icons.Default.Favorite,
                    contentDescription = null,
                    tint = Color(0xFFEEEEEE)
                )
            }
        }
    }
}

@Composable
fun BottomNavigationBar(modifier: Modifier = Modifier) {
    Card(
        modifier = modifier
            .fillMaxWidth()
            .padding(24.dp),
        shape = RoundedCornerShape(32.dp),
        elevation = CardDefaults.cardElevation(defaultElevation = 8.dp),
        colors = CardDefaults.cardColors(containerColor = Color(0xFF1E1E1E))
    ) {
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .padding(vertical = 16.dp),
            horizontalArrangement = Arrangement.SpaceEvenly
        ) {
            Icon(Icons.Default.Home, contentDescription = null, tint = Color.White)
            Icon(Icons.Default.Search, contentDescription = null, tint = Color.Gray)
            Icon(Icons.Default.Favorite, contentDescription = null, tint = Color.Gray)
            Icon(Icons.Default.Face, contentDescription = null, tint = Color.Gray)
        }
    }
}

data class Course(
    val title: String,
    val subtitle: String,
    val rating: String,
    val bgColor: Color,
    val accentColor: Color
)

@Preview(showBackground = true)
@Composable
fun DefaultPreview() {
    AppTheme {
        CourseDashboardScreen()
    }
}

File: suc/swift/078.swift
import SwiftUI

@main
struct CourseApp: App {
    var body: some Scene {
        WindowGroup {
            ContentView()
        }
    }
}

struct ContentView: View {
    var body: some View {
        CourseDashboardScreen()
            .preferredColorScheme(.light)
            .statusBarHidden(true)
    }
}

// MARK: - Models
struct Course: Identifiable {
    let id = UUID()
    let title: String
    let subtitle: String
    let rating: String
    let bgColor: Color
    let accentColor: Color
    let iconName: String
}

// MARK: - Theme
struct AppTheme {
    static let primary = Color(red: 0.31, green: 0.19, blue: 0.67) // 0xFF4E31AA
    static let secondary = Color(red: 0.23, green: 0.11, blue: 0.44) // 0xFF3A1C71
    static let background = Color(red: 0.96, green: 0.96, blue: 0.97) // 0xFFF5F5F7
    static let surface = Color.white
    static let onBackground = Color(red: 0.11, green: 0.11, blue: 0.12) // 0xFF1C1B1F
    static let progressColor = Color(red: 1.0, green: 0.65, blue: 0.15) // 0xFFFFA726
    static let selectedCategoryColor = Color(red: 1.0, green: 0.44, blue: 0.26) // 0xFFFF7043
}

// MARK: - Main Screen
struct CourseDashboardScreen: View {
    var body: some View {
        ZStack {
            AppTheme.background.ignoresSafeArea()
            
            VStack(spacing: 0) {
                ScrollView(showsIndicators: false) {
                    VStack(spacing: 0) {
                        HeaderSection()
                        SearchBarSection()
                        OngoingCourseSection()
                        CategoriesSection()
                        RecommendedSection()
                        Spacer().frame(height: 100) // Bottom padding for floating nav bar
                    }
                }
            }
            
            VStack {
                Spacer()
                BottomNavigationBar()
                    .padding(.bottom, 24)
            }
        }
        .edgesIgnoringSafeArea(.bottom)
    }
}

// MARK: - Sections

struct HeaderSection: View {
    var body: some View {
        HStack {
            VStack(alignment: .leading, spacing: 4) {
                Text("Hello, Sophie")
                    .font(.system(size: 28, weight: .bold)) // HeadlineMedium
                    .foregroundColor(AppTheme.onBackground)
                Text("Let's start learning!")
                    .font(.system(size: 16)) // BodyLarge
                    .foregroundColor(.gray)
            }
            
            Spacer()
            
            HStack(spacing: 8) {
                Button(action: {}) {
                    Image(systemName: "bell.fill")
                        .font(.system(size: 24))
                        .foregroundColor(AppTheme.onBackground)
                }
                
                ZStack {
                    Circle()
                        .fill(AppTheme.primary.opacity(0.2))
                        .frame(width: 48, height: 48)
                    
                    Image(systemName: "face.smiling.fill") // Replaced Face
                        .font(.system(size: 24))
                        .foregroundColor(AppTheme.primary)
                }
            }
        }
        .padding(24)
    }
}

struct SearchBarSection: View {
    var body: some View {
        HStack {
            Image(systemName: "magnifyingglass")
                .foregroundColor(.gray)
            
            Spacer().frame(width: 16)
            
            Text("Search for course...")
                .foregroundColor(Color(UIColor.lightGray))
            
            Spacer()
            
            Image(systemName: "slider.horizontal.3") // Replaced Settings icon for filter
                .foregroundColor(AppTheme.primary)
        }
        .padding(.horizontal, 16)
        .frame(height: 56)
        .background(Color.white)
        .cornerRadius(16)
        .padding(.horizontal, 24)
    }
}

struct OngoingCourseSection: View {
    var body: some View {
        VStack(alignment: .leading, spacing: 16) {
            Text("Ongoing Course")
                .font(.system(size: 20, weight: .bold)) // TitleLarge
                .foregroundColor(AppTheme.onBackground)
            
            ZStack {
                RoundedRectangle(cornerRadius: 24)
                    .fill(AppTheme.primary)
                    .frame(height: 140)
                
                HStack(spacing: 16) {
                    VStack(alignment: .leading) {
                        Text("3D Modeling")
                            .font(.system(size: 14))
                            .foregroundColor(.white.opacity(0.7))
                        
                        Spacer().frame(height: 8)
                        
                        Text("Blender Basic")
                            .font(.system(size: 22, weight: .bold))
                            .foregroundColor(.white)
                        
                        Spacer().frame(height: 16)
                        
                        HStack(alignment: .center) {
                            Text("75%")
                                .font(.system(size: 12))
                                .foregroundColor(.white)
                            
                            Spacer().frame(width: 8)
                            
                            GeometryReader { geometry in
                                ZStack(alignment: .leading) {
                                    RoundedRectangle(cornerRadius: 4)
                                        .fill(Color.white.opacity(0.2))
                                        .frame(height: 6)
                                    
                                    RoundedRectangle(cornerRadius: 4)
                                        .fill(AppTheme.progressColor)
                                        .frame(width: geometry.size.width * 0.75, height: 6)
                                }
                            }
                            .frame(height: 6)
                        }
                    }
                    
                    Spacer() // Spacer in HStack to push play button to right
                    
                    ZStack {
                        Circle()
                            .fill(Color.white)
                            .frame(width: 60, height: 60)
                        
                        Image(systemName: "play.fill")
                            .font(.system(size: 32)) // Adjusted size
                            .foregroundColor(AppTheme.primary)
                            .offset(x: 2) // Visual center adjustment
                    }
                }
                .padding(20)
            }
        }
        .padding(24)
    }
}

struct CategoriesSection: View {
    let categories = ["Coding", "Design", "Business", "Photo", "Music"]
    
    var body: some View {
        ScrollView(.horizontal, showsIndicators: false) {
            HStack(spacing: 12) {
                // Leading padding
                Spacer().frame(width: 12)
                
                ForEach(categories, id: \.self) { category in
                    let isSelected = category == "Design"
                    Text(category)
                        .font(.system(size: 14, weight: .medium))
                        .foregroundColor(isSelected ? .white : .gray)
                        .padding(.horizontal, 24)
                        .padding(.vertical, 12)
                        .background(isSelected ? AppTheme.selectedCategoryColor : Color.white)
                        .cornerRadius(12)
                }
                
                // Trailing padding
                Spacer().frame(width: 12)
            }
        }
    }
}

struct RecommendedSection: View {
    let courses = [
        Course(title: "Mobile Dev", subtitle: "Android Compose", rating: "4.9", bgColor: Color(red: 0.89, green: 0.95, blue: 0.99), accentColor: Color(red: 0.08, green: 0.40, blue: 0.75), iconName: "checkmark.circle.fill"),
        Course(title: "Photography", subtitle: "Master Lighting", rating: "4.7", bgColor: Color(red: 0.99, green: 0.89, blue: 0.93), accentColor: Color(red: 0.76, green: 0.09, blue: 0.36), iconName: "star.fill"),
        Course(title: "Finance", subtitle: "Crypto Trading", rating: "4.5", bgColor: Color(red: 0.91, green: 0.96, blue: 0.92), accentColor: Color(red: 0.18, green: 0.49, blue: 0.20), iconName: "star.fill")
    ]
    
    var body: some View {
        VStack(alignment: .leading, spacing: 16) {
            HStack {
                Text("Recommended")
                    .font(.system(size: 20, weight: .bold)) // TitleLarge
                    .foregroundColor(AppTheme.onBackground)
                Spacer()
                Text("See all")
                    .foregroundColor(.gray)
            }
            
            ForEach(courses) { course in
                CourseCard(course: course)
            }
        }
        .padding(24)
    }
}

struct CourseCard: View {
    let course: Course
    
    var body: some View {
        ZStack {
            RoundedRectangle(cornerRadius: 20)
                .fill(Color.white)
                .frame(height: 100)
            
            HStack(spacing: 0) {
                // Left Icon Box
                ZStack {
                    course.bgColor
                    Image(systemName: course.iconName)
                        .font(.system(size: 32))
                        .foregroundColor(course.accentColor)
                }
                .frame(width: 100)
                .clipShape(CustomCornerShape(radius: 20, corners: [.topLeft, .bottomLeft]))
                
                // Content
                VStack(alignment: .leading, spacing: 4) {
                    Text(course.subtitle)
                        .font(.system(size: 12, weight: .bold))
                        .foregroundColor(AppTheme.primary)
                    
                    Text(course.title)
                        .font(.system(size: 16, weight: .bold)) // TitleMedium
                        .foregroundColor(AppTheme.onBackground)
                    
                    HStack(spacing: 4) {
                        Image(systemName: "star.fill")
                            .font(.system(size: 12)) // Adjusted size
                            .foregroundColor(Color(red: 1.0, green: 0.70, blue: 0.0)) // Amber
                        
                        Text(course.rating)
                            .font(.system(size: 12))
                            .foregroundColor(.gray)
                    }
                }
                .padding(16)
                
                Spacer()
                
                // Favorite Icon
                Button(action: {}) {
                    Image(systemName: "heart.fill")
                        .foregroundColor(Color(UIColor.systemGray5)) // #EEEEEE equivalent usually very light gray
                }
                .padding(.trailing, 16)
            }
        }
        .frame(height: 100)
    }
}

struct BottomNavigationBar: View {
    var body: some View {
        HStack(alignment: .center) {
            Spacer()
            Image(systemName: "house.fill")
                .foregroundColor(.white)
            Spacer()
            Image(systemName: "magnifyingglass")
                .foregroundColor(.gray)
            Spacer()
            Image(systemName: "heart.fill")
                .foregroundColor(.gray)
            Spacer()
            Image(systemName: "person.fill")
                .foregroundColor(.gray)
            Spacer()
        }
        .padding(.vertical, 16)
        .background(Color(red: 0.12, green: 0.12, blue: 0.12)) // #1E1E1E
        .cornerRadius(32)
        .shadow(color: Color.black.opacity(0.2), radius: 8, x: 0, y: 4)
        .padding(.horizontal, 24)
    }
}

// MARK: - Helper Shapes
struct CustomCornerShape: Shape {
    var radius: CGFloat
    var corners: UIRectCorner
    
    func path(in rect: CGRect) -> Path {
        let path = UIBezierPath(
            roundedRect: rect,
            byRoundingCorners: corners,
            cornerRadii: CGSize(width: radius, height: radius)
        )
        return Path(path.cgPath)
    }
}